### PR TITLE
Add means of getting the parameter associated with a handle 

### PR DESCRIPTION
**Pull request by hartytp**
This PR adds a helper method to get the parameter associated with a given handle. Changes in the associated parameter are tracked across rebinds.

closes https://github.com/OxfordIonTrapGroup/ndscan/issues/316

### CODE REVIEW

File: ndscan/experiment/parameters.py [hunk 1]
```diff
@@ -213,18 +215,23 @@ class ParamHandle:
...
         assert name.isidentifier(), ("ParamHandle name should be the identifier it is "
                                      "referred to as in the owning fragment.")
 
         self._store = None
         self._changed_after_use = True
 
+    def get_parameter(self) -> "ParamBase":
+        """Returns the parameter associated with this handle."""
+        return self._parameter
+
```
Comment (dnadlinger) [inline ndscan/experiment/parameters.py:234]: Can just be a public field like `owner`/`name`. Python supports properties anyway if this ever needs to change (or we really wanted to enforce read-only-ness).

**Comment by dnadlinger** [comment]
Regarding tracking rebinds, I guess you have some intuition about why this is the "correct" behaviour? Doesn't seem obvious to me either way.

**Comment by dnadlinger** [comment]
I'm happy to merge the unrelated small chagnes as part of this PR for simplicity, but could you rebase to merge the two first commits (to keep the unit tests passing on all commits)?

**Comment by hartytp** [comment]
> Regarding tracking rebinds, I guess you have some intuition about why this is the "correct" behaviour? Doesn't seem obvious to me (but neither would the converse).

Good question.

My assumption was "If I ask a parameter handle what metadata is associated with it the answer I get should reflect what is actually going to happen when I run the experiment, rather than how the fragment was originally created." and "the metadata I get should not depend on which handle I ask".

Maybe there is a different intuition I'm missing, but those just seemed to be reasonable assumptions. Either way, I think this needs calling out explicitly in the documentation so once we've concluded this I'll update that.

> This might also limit the usefulness of this in general/require careful documentation, as the semantics may not be that obvious, and accessing the parameter metadata is almost never necessary in user code. Omitting this was not a principled decision either, but just made to keep the API surface minimal initially (in terms of pointing users in the right direction).

I've found a number of times when accessing parameter metadata has been useful in user code. For example, suppose I want to make a small framework which automatically pushes analysis results to the dataset linked to a parameter.

File: ndscan/experiment/parameters.py (hunk 1 above)
Comment (hartytp) [reply ndscan/experiment/parameters.py:234]: Happy to do that if you prefer

**Comment by hartytp** [comment]
> I'm happy to merge the unrelated small chagnes as part of this PR for simplicity, but could you rebase to merge the two first commits (to keep the unit tests passing on all commits)?

Sorry, I'm not sure I follow the request here. Do you mean essentially rewriting history so I fix the tests before adding the new feature?

**Comment by dnadlinger** [comment]
> Sorry, I'm not sure I follow the request here. Do you mean essentially rewriting history so I fix the tests before adding the new feature?

Just `rebase -i` to merge 27ebbd2 and 42c7372, as either alone would lead to the CI breaking. (Not that we need to bisect the history often for ndscan to track down some bugs, but gratuitous test failures can be annoying in that case.)

**Comment by dnadlinger** [comment]
Great, thanks. Could you just remove the `get_parameter` accessor as well, and add one very basic assertion to the parameter tests checking that `parameter` is set to the value passed to the constructor, so we have a regression test for the existence of the API?

Happy to merge this after that and a short docstring for `ParamHandle.parameter`.

**Comment by hartytp** [comment]
Done - will merge once CI passes if you're happy. Thanks for the quick review!

**Comment by dnadlinger** [comment]
Huh, I was trying to check whether the `.parameters` attribute shows up as being documented (you might need to add a `#: foo` docstring before the `self.parameter = …` line), but it seems like all the `parameters` classes are missing from the API docs.

Could you rebase again please to merge the fixups into the first commit (and only leave the unrelated changes separately)?

**Comment by dnadlinger** [comment]
Oh, it's probably the __all__…

**Comment by dnadlinger** [comment]
Ah, the branch is on the OITG fork anyway, so I'll just fix there myself.

**Comment by dnadlinger** [comment]
Thanks!

**Comment by hartytp** [comment]
@dnadlinger not sure if useful, but I've found the workflows at `ionics_fits` for docs relatively useful - e.g. having a doc build as a test helps to keep the quality up (also a `poe` task is quite useful when checking things locally). I'd be happy to submit a PR if you ever wanted anything along those lines added

**Comment by hartytp** [comment]
>Doc builds are already run as part of CI (and uploaded to https://oxfordiontrapgroup.github.io/ndscan/apidocs.html#module-ndscan.experiment.parameters for master commits).

aah, cool, I must have missed how this works. Does the current workflow prevent merging PRs with doc build errors (e.g. missing references)?

**Comment by dnadlinger** [comment]
The doc builds are marked as "requried", so if `make html` outright fails, it will indeed block merges. Sphinx warnings probably don't currently count as errors right now.

**Comment by dnadlinger** [comment]
Added a `docs` target in https://github.com/OxfordIonTrapGroup/ndscan/commit/6a229ee664a029d1b61d9079f1a5afe06f8175f8.

**Pull request merged**